### PR TITLE
refactored to handle registerDeferredDeepLinkHandler before startSession

### PR DESCRIPTION
**Pull request by anismiles**

### CODE REVIEW

File: mParticle-Singular/MPKitSingular.m [hunk 1]
```diff
@@ -68,6 +74,11 @@ - (void)start{
         /*
          Start your SDK here. The configuration dictionary can be retrieved from self.configuration
          */
+        ddlLink = nil;
+        [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
+            ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,SINGULAR_DEEPLINK_KEY, nil];
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m]: @anismiles you should be invoking the deep link completion handler here, if not nil

File: mParticle-Singular/MPKitSingular.m [hunk 2]
```diff
@@ -8,6 +8,12 @@
 // This is temporary to allow compilation (will be provided by core SDK)
 NSUInteger MPKitInstanceSingularTemp = 119;
 
+@interface MPKitSingular() {
+    NSDictionary *ddlLink;
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m:12]: Can you also add a property for the deep link completion handler that's passed into the kit from the implementing app?

File: mParticle-Singular/MPKitSingular.m [hunk 3]
```diff
@@ -98,21 +108,18 @@ If your company SDK instance is available and is applicable (Please return nil i
     }
 }
 
-
 #pragma mark Application
 - (MPKitExecStatus *)checkForDeferredDeepLinkWithCompletionHandler:(void(^)(NSDictionary *linkInfo, NSError *error))completionHandler {
-    [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
-        NSDictionary *ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,SINGULAR_DEEPLINK_KEY, nil];
+    if (_started && (ddlLink)){
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m:117]: if neither of these are true, you should be saving a reference to the completion handler for later use.

File: mParticle-Singular/MPKitSingular.m [hunk 4]
```diff
@@ -123,6 +130,11 @@ - (MPKitExecStatus *)setDeviceToken:(NSData *)deviceToken {
 - (nonnull MPKitExecStatus *)continueUserActivity:(nonnull NSUserActivity *)userActivity restorationHandler:(void(^ _Nonnull)(NSArray * _Nullable restorableObjects))restorationHandler {
     
     if([userActivity.activityType isEqualToString:NSUserActivityTypeBrowsingWeb]){
+        ddlLink = nil;
+        [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
+            ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,SINGULAR_DEEPLINK_KEY, nil];
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m]: similar to during kit initialization, you should be invoking the completion handler property here.

File: mParticle-Singular/MPKitSingular.m [hunk 5]
```diff
@@ -216,13 +227,22 @@ - (MPKitExecStatus *)logCommerceEvent:(MPCommerceEvent *)commerceEvent {
...
+    }
     MPKitExecStatus *execStatus = [[MPKitExecStatus alloc] initWithSDKCode:@(MPKitInstanceSingularTemp) returnCode:MPKitReturnCodeSuccess];
     return execStatus;
 }
 
 - (nonnull MPKitExecStatus *)openURL:(nonnull NSURL *)url sourceApplication:(nullable NSString *)sourceApplication annotation:(nullable id)annotation {
     if(url){
+        ddlLink = nil;
+        [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
+            ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,SINGULAR_DEEPLINK_KEY, nil];
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m]: same comment as above regarding completion handler

File: mParticle-Singular/MPKitSingular.m [hunk 6]
```diff
@@ -231,6 +251,11 @@ - (nonnull MPKitExecStatus *)openURL:(nonnull NSURL *)url sourceApplication:(nul
 
 - (nonnull MPKitExecStatus *)openURL:(nonnull NSURL *)url options:(nullable NSDictionary<NSString *, id> *)options {
     if(url){
+        ddlLink = nil;
+        [Singular registerDeferredDeepLinkHandler:^(NSString *deeplink) {
+            ddlLink = [[NSDictionary alloc] initWithObjectsAndKeys:deeplink,SINGULAR_DEEPLINK_KEY, nil];
```
Comment (samdozor) [inline mParticle-Singular/MPKitSingular.m]: same comment as above.

**Review by samdozor — CHANGES_REQUESTED** [review]

File: mParticle-Singular/MPKitSingular.m (hunk 2 above)
Comment (anismiles) [reply mParticle-Singular/MPKitSingular.m:12]: Done.

File: mParticle-Singular/MPKitSingular.m (hunk 3 above)
Comment (anismiles) [reply mParticle-Singular/MPKitSingular.m:117]: Done.

**Review by samdozor — APPROVED** [review]
LGTM! Thanks @anismiles

**Pull request merged**